### ISSUE TITLE
JSON API security

### ISSUE THREAD

**Issue by fabianfrz** [opnsense/core#2672]
It may be a good idea to discuss if this is really a good idea:

https://github.com/opnsense/core/blob/master/src/opnsense/mvc/app/controllers/OPNsense/Base/ApiControllerBase.php#L217

The problem here is that this changes the data for all clients, not only the web configuration. The representation code should be responsible to bring it to the format required by the viewer. In our case probably in the `mapDataToFormUI` function.

**Comment by AdSchellevis** [comment]
I haven't seen any real life issues with it, it has been in there for quite some time.  Originally we added the escaping here https://github.com/opnsense/core/commit/3c2aad981f6bf5fe513ca9a276b1f11133f18b47

Moving it to the page might be an option, although the impact will likely be more then `mapDataToFormUI()`.

**Comment by fabianfrz** [comment]
yes because it will probably impact code, that is not part of OPNsense (external API clients like monitoring solutions, configuration viewer etc.). For example, if I would write a viewer in C++/QT, I would very likely get an &amp;amp; in the text field unless I revert this change.

**Comment by AdSchellevis** [comment]
In theory I agree, although we need a solid option on the client side to prevent (security) issues. It's not very high on my list of things todo at the moment, and changing `mapDataToFormUI()` will very likely not be enough to prevent xss issues in the current codebase.

**Comment by AdSchellevis** [comment]
Thinking about this again, it's only an issue for external clients, so let's treat it like that.

Global escaping our data which is heading for our webclient is actually desired behaviour, since it will make all of our lives a lot harder if we need to validate all api calls in the browser the way it is now. 

In theory it would be better to solve this client side, but since we don't have the same level of control there, the pragmatical option is to make sure we only send escaped data to the client. 

I'll refactor the code to send raw (unescaped) data to external clients, and schedule this for inclusion in 19.1.

**Comment by fichtner** [comment]
After some thought and review I agree with this approach. Good idea. 👍

**Comment by AdSchellevis** [comment]
@fichtner thanks
